File: Suspension matlab/Damping/Critical_Damping_Analysis/Critical_Damping_Analysis.m
%% Critical Damping Analysis
clear; clc; close all;
scriptInfo = dbstack('-completenames');
scriptDir = fileparts(scriptInfo(1).file);
addpath(scriptDir);
params = load_data(fullfile(scriptDir,'data.csv'));
paramNames = fieldnames(params);
for paramIdx = 1:numel(paramNames)
    eval([paramNames{paramIdx} ' = params.(paramNames{paramIdx});']);
end

%% Parameters

m = critical_m;      % sprung mass (kg)

%% Spring stiffness range

k = critical_k;

%% Critical damping

cc = 2 * sqrt(k .* m);

%% Natural frequency

fn = (1/(2*pi)) * sqrt(k./m);

%% Plot

figure;

subplot(2,1,1)

plot(k/1000,fn,'LineWidth',2);

grid on;

xlabel('Spring Stiffness (kN/m)');
ylabel('Natural Frequency (Hz)');

title('Natural Frequency vs Spring Stiffness');

subplot(2,1,2)

plot(k/1000,cc,'LineWidth',2);

grid on;

xlabel('Spring Stiffness (kN/m)');
ylabel('Critical Damping (N·s/m)');

title('Critical Damping vs Spring Stiffness');

File: Suspension matlab/Damping/Critical_Damping_Analysis/Critical_Damping_Analysis.py
from pathlib import Path
import sys

import matplotlib.pyplot as plt
import numpy as np

ROOT = next(p for p in Path(__file__).resolve().parents if (p / "suspension_utils.py").exists())
sys.path.insert(0, str(ROOT))

from suspension_utils import load_data, show_plots


def main() -> None:
    p = load_data(Path(__file__).with_name("data.csv"))
    m = p["critical_m"]
    k = np.asarray(p["critical_k"], dtype=float)

    cc = 2 * np.sqrt(k * m)
    fn = (1 / (2 * np.pi)) * np.sqrt(k / m)

    fig, axes = plt.subplots(2, 1)
    axes[0].plot(k / 1000, fn, linewidth=2)
    axes[0].grid(True)
    axes[0].set_xlabel("Spring Stiffness (kN/m)")
    axes[0].set_ylabel("Natural Frequency (Hz)")
    axes[0].set_title("Natural Frequency vs Spring Stiffness")
    axes[1].plot(k / 1000, cc, linewidth=2)
    axes[1].grid(True)
    axes[1].set_xlabel("Spring Stiffness (kN/m)")
    axes[1].set_ylabel("Critical Damping (N*s/m)")
    axes[1].set_title("Critical Damping vs Spring Stiffness")
    show_plots()


if __name__ == "__main__":
    main()
